File: web-app/backend/app/routers/copilot.py
"""Copilot router — 4 endpoints driving /copilot in the frontend.

  GET    /copilot/conversations              -- list (recent first)
  POST   /copilot/conversations              -- create empty thread
  GET    /copilot/conversations/{id}         -- header + full message history
  POST   /copilot/conversations/{id}/messages -- send user msg, SSE stream

Authentication: same JWT bearer as /buildings and /portfolio.
"""
from __future__ import annotations

from typing import Annotated
from uuid import UUID, uuid4

import jwt
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import StreamingResponse
from fastapi.security import HTTPAuthorizationCredentials
from sqlalchemy.orm import Session

from app.db.database import get_db
from app.repositories import copilot as copilot_repo
from app.schemas.copilot import (
    ConversationCreatedResponse,
    ConversationCreateRequest,
    ConversationDetail,
    ConversationItem,
    ConversationListResponse,
    MessageCreateRequest,
    MessageItem,
)
from app.services.copilot.orchestrator import sse_event, stream_assistant_response
from app.utils.jwt import (
    bearer_scheme,
    decode_access_token,
    get_current_user_id,
)

router = APIRouter(prefix="/copilot", tags=["copilot"])


def _get_user_and_org(
    user_id: Annotated[UUID, Depends(get_current_user_id)],
    credentials: Annotated[HTTPAuthorizationCredentials | None, Depends(bearer_scheme)],
) -> tuple[UUID, UUID]:
    """Pull user_id + organization_id out of the JWT.

    The JWT issued by /auth/login already carries org_id (the user's personal
    workspace at signup). We trust it rather than re-querying Postgres on
    every request — same pattern as get_current_user_id.
    """
    if credentials is None:
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, "Missing Authorization header")
    claims = decode_access_token(credentials.credentials)
    try:
        org_id = UUID(claims["org_id"])
    except (KeyError, ValueError):
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, "Token missing valid 'org_id' claim")
    return user_id, org_id


# ----- GET /conversations ----------------------------------------------


@router.get("/conversations", response_model=ConversationListResponse)
def list_conversations(
    user_and_org: Annotated[tuple[UUID, UUID], Depends(_get_user_and_org)],
    db: Annotated[Session, Depends(get_db)],
    limit: int = 50,
    include_archived: bool = False,
) -> ConversationListResponse:
    user_id, _ = user_and_org
    rows = copilot_repo.list_conversations_for_user(
        db,
        user_id=user_id,
        limit=limit,
        include_archived=include_archived,
    )
    return ConversationListResponse(
        conversations=[ConversationItem(**r) for r in rows]
    )


# ----- POST /conversations ---------------------------------------------


@router.post(
    "/conversations",
    response_model=ConversationCreatedResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_conversation(
    body: ConversationCreateRequest,
    user_and_org: Annotated[tuple[UUID, UUID], Depends(_get_user_and_org)],
    db: Annotated[Session, Depends(get_db)],
) -> ConversationCreatedResponse:
    user_id, org_id = user_and_org

    # Title: caller-provided OR a placeholder. Auto-titling from the first
    # user message is a V1.5 polish (extra Claude call after first turn).
    title = (body.title or "New Chat").strip()[:255]

    conv = copilot_repo.create_conversation(
        db,
        user_id=user_id,
        organization_id=org_id,
        title=title,
        fabric_building_id=body.fabric_building_id,
    )
    return ConversationCreatedResponse(id=conv.id, title=conv.title)


# ----- GET /conversations/{id} -----------------------------------------


@router.get("/conversations/{conversation_id}", response_model=ConversationDetail)
def get_conversation(
    conversation_id: UUID,
    user_and_org: Annotated[tuple[UUID, UUID], Depends(_get_user_and_org)],
    db: Annotated[Session, Depends(get_db)],
) -> ConversationDetail:
    user_id, _ = user_and_org
    looked_up = copilot_repo.get_conversation_with_building(
        db, conversation_id=conversation_id, user_id=user_id,
    )
    if looked_up is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Conversation not found.")
    conv, fab_id = looked_up
    messages = copilot_repo.list_messages(db, conversation_id=conversation_id)
    return ConversationDetail(
        id=conv.id,
        title=conv.title,
        provider=conv.provider,
        model=conv.model,
        fabric_building_id=fab_id,
        is_archived=conv.is_archived,
        last_message_at=conv.last_message_at,
        created_at=conv.created_at,
        updated_at=conv.updated_at,
        messages=[MessageItem.model_validate(m) for m in messages],
    )


# ----- POST /conversations/{id}/messages  (SSE) ------------------------


@router.post("/conversations/{conversation_id}/messages")
async def post_message(
    conversation_id: UUID,
    body: MessageCreateRequest,
    user_and_org: Annotated[tuple[UUID, UUID], Depends(_get_user_and_org)],
    db: Annotated[Session, Depends(get_db)],
) -> StreamingResponse:
    """Send a user message and stream the assistant's response via SSE.

    Response content-type: text/event-stream. Events:

      event: user_message_persisted    -- {"id": "<uuid>"}
      event: text                      -- {"text": "<chunk>"}
      event: tool_call_start           -- {"tool_use_id", "name"}
      event: tool_call                 -- {"tool", "tool_use_id", "input"}
      event: tool_result               -- {"tool", "tool_use_id", "result_preview", "is_error"}
      event: complete                  -- {"message_id", "stop_reason", "tokens_input", "tokens_output"}
      event: error                     -- {"message"}
    """
    user_id, org_id = user_and_org

    # Verify access before opening the stream — return a clean 404 instead
    # of starting the SSE and emitting an error event (less surprising).
    looked_up = copilot_repo.get_conversation_with_building(
        db, conversation_id=conversation_id, user_id=user_id,
    )
    if looked_up is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Conversation not found.")

    async def event_stream():
        try:
            async for chunk in stream_assistant_response(
                db,
                conversation_id=conversation_id,
                user_id=user_id,
                organization_id=org_id,
                user_message=body.content,
            ):
                yield chunk
        except Exception as exc:  # noqa: BLE001 — last-resort safety
            yield sse_event("error", {
                "message": f"Stream aborted: {type(exc).__name__}: {exc}",
            })

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache, no-transform",
            "X-Accel-Buffering": "no",  # disable nginx/proxy buffering
            "Connection": "keep-alive",
        },
    )

File: web-app/frontend/lib/api/copilot.ts
/**
 * Types + server-only fetch helpers for /copilot endpoints.
 *
 * Stays in lockstep with backend app/schemas/copilot.py.
 *
 * Architecture:
 *   - Initial conversation list: server component fetches via fetchConversations()
 *     (same pattern as lib/api/portfolio.ts).
 *   - Create conversation + send message: client components call the
 *     /api/copilot/* proxy routes (Next.js route handlers forward to backend
 *     with the user's JWT). Keeps the access token off the client.
 *   - SSE streaming: see lib/copilot/sseClient.ts — also goes through the
 *     proxy so the backend URL never appears in browser network logs.
 */

// ============================================================================
// Wire types — match backend app/schemas/copilot.py
// ============================================================================

export type MessageRole = "user" | "assistant" | "tool"

export type ToolCall = {
  id: string
  name: string
  input: Record<string, unknown>
}

export type MessageItem = {
  id: string
  role: MessageRole
  content: string | null
  tool_calls: ToolCall[] | ToolCall | null
  tool_call_id: string | null
  tool_name: string | null
  tokens_input: number | null
  tokens_output: number | null
  latency_ms: number | null
  created_at: string // ISO datetime
}

export type ConversationItem = {
  id: string
  title: string
  provider: string
  model: string | null
  fabric_building_id: string | null
  is_archived: boolean
  last_message_at: string | null
  created_at: string
  updated_at: string
}

export type ConversationListResponse = {
  conversations: ConversationItem[]
}

export type ConversationDetail = ConversationItem & {
  messages: MessageItem[]
}

export type ConversationCreatedResponse = {
  id: string
  title: string
}

// ============================================================================
// SSE event shapes (frontend-facing — backend orchestrator emits these)
// ============================================================================

export type SSEEvent =
  | { type: "user_message_persisted"; data: { id: string } }
  | { type: "text"; data: { text: string } }
  | { type: "tool_call_start"; data: { tool_use_id: string; name: string } }
  | {
      type: "tool_call"
      data: { tool: string; tool_use_id: string; input: Record<string, unknown> }
    }
  | {
      type: "tool_result"
      data: {
        tool: string
        tool_use_id: string
        result_preview: string
        is_error: boolean
      }
    }
  | {
      type: "complete"
      data: {
        message_id: string | null
        stop_reason: string
        tokens_input?: number
        tokens_output?: number
        latency_ms?: number
        warning?: string
      }
    }
  | { type: "error"; data: { message: string } }

// ============================================================================
// Server-only helpers
// ============================================================================

type FetchResult<T> = { ok: true; data: T } | { ok: false; error: string }

function getBackendUrl(): string | null {
  return process.env.BACKEND_URL || null
}

function logError(scope: string, status: number, body: string) {
  console.error(`[${scope}] Backend ${status}: ${body.slice(0, 200)}`)
}

/**
 * GET /copilot/conversations — list user's recent conversations.
 *
 * Server-side only. Pass session.accessToken from a server component.
 */
export async function fetchConversations(
  accessToken: string
): Promise<FetchResult<ConversationListResponse>> {
  const backendUrl = getBackendUrl()
  if (!backendUrl) {
    console.error("[fetchConversations] BACKEND_URL not configured")
    return { ok: false, error: "Server misconfigured" }
  }
  if (!accessToken) {
    return { ok: false, error: "Missing access token" }
  }

  try {
    const res = await fetch(`${backendUrl}/copilot/conversations`, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        Accept: "application/json",
      },
      cache: "no-store",
    })

    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      logError("fetchConversations", res.status, body)
      return { ok: false, error: `Backend ${res.status}` }
    }

    const data = (await res.json()) as ConversationListResponse
    return { ok: true, data }
  } catch (err) {
    console.error("[fetchConversations] Network/unknown error:", err)
    return { ok: false, error: "Network error" }
  }
}

/**
 * GET /copilot/conversations/{id} — full conversation with message history.
 *
 * Returns ok:false with error="Not found" on 404 (conversation doesn't
 * exist OR user not authorized — backend conflates them).
 */
export async function fetchConversation(
  accessToken: string,
  conversationId: string
): Promise<FetchResult<ConversationDetail>> {
  const backendUrl = getBackendUrl()
  if (!backendUrl) {
    console.error("[fetchConversation] BACKEND_URL not configured")
    return { ok: false, error: "Server misconfigured" }
  }
  if (!accessToken) {
    return { ok: false, error: "Missing access token" }
  }

  try {
    const res = await fetch(
      `${backendUrl}/copilot/conversations/${encodeURIComponent(conversationId)}`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${accessToken}`,
          Accept: "application/json",
        },
        cache: "no-store",
      }
    )

    if (res.status === 404) {
      return { ok: false, error: "Not found" }
    }

    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      logError("fetchConversation", res.status, body)
      return { ok: false, error: `Backend ${res.status}` }
    }

    const data = (await res.json()) as ConversationDetail
    return { ok: true, data }
  } catch (err) {
    console.error("[fetchConversation] Network/unknown error:", err)
    return { ok: false, error: "Network error" }
  }
}
